add options to require this/Me for fields, properties, methods, and events

brettfo opened the pull request:
This is a re-creation of PR #7296 to address feedback.  Since there are so many changes I opted to close the other PR and open a new one to preserve an easy comment history and enable a side-by-side diff.

### Description
Splits the option _"Qualify member access with 'this' or 'Me'"_ into four separate options (and marks the former as `[Obsolete]`):

- Qualify **field** access.
- Qualify **property** access.
- Qualify **method** access.
- Qualify **event** access.

### Known Issues
1. We've discussed giving the user a dial in how and when these properties should be applied, e.g., "only apply this style on newly generated code" all the way up to "flag violations of this style as a build break" and various steps in between.  That has not been addressed in this PR and is still an open discussion item that I filed as #7289.
2. If a user currently has _"Qualify member access with 'this' or 'Me'"_ set to `true`, we need some way to upgrade/migrate that setting to the 4 new separate settings.

Fixes #7065.
Fixes #7090.
Also fixes an unrelated bug I found in Encapsulate Field in VB that caused `Me.` to be improperly included when an escaped keyword is used when the option wasn't set and also enables Encapsulate Field in C# for escaped identifiers.

brettfo commented:
test vsi please

brettfo commented:
test vsi please

brettfo commented:
The failing vsi p3 queue is expected and due to the `OptionsDTE.xml` test which I have updated in the internal PR dotnet/roslyn-internal#430.

brettfo commented:
Tagging @dotnet/roslyn-ide for review.

brettfo commented:
To make reviewing this easier, some interesting parts are as follows:
- The fix to the simplifier is here: [C#](https://github.com/dotnet/roslyn/pull/7540/files#diff-a419a7e5e40e4a95b7469c2a47229a56R702), [VB](https://github.com/dotnet/roslyn/pull/7540/files#diff-b356e6af8a9080a75b0869d50e2876e4R952).
- The common code to determine if `this.`/`Me.` should get removed or not is [here](https://github.com/dotnet/roslyn/pull/7540/files#diff-6cba5e83d3f2861af2117e7fa7fcfdebR103).
- The tests for the added diagnostic/code fixer are in [`QualifyMemberAccessTests.cs`](https://github.com/dotnet/roslyn/pull/7540/files#diff-f99c608e247a37300d9896c720fbe6baR1) and [`QualifyMemberAccessTests.vb`](https://github.com/dotnet/roslyn/pull/7540/files#diff-4cc78320b838c44bc5698e31cffd7f6dR1).
- The diagnostic analyzers are here: [C#](https://github.com/dotnet/roslyn/pull/7540/files#diff-142cd7ea60e5c42d11095674aaebd74bR1), [VB](https://github.com/dotnet/roslyn/pull/7540/files#diff-2cb3e5983574aa4d8e22722ad44a1f47R1), [abstract base class](https://github.com/dotnet/roslyn/pull/7540/files#diff-799060ac52680fe6fd3b830824734355R1).  Note that the abstract base class also calls into the shared helper for determining if `this.`/`Me.` should be removed.
- The code fix providers are here: [C#](https://github.com/dotnet/roslyn/pull/7540/files#diff-cbeb90267ef67343ab6910e820008769R1), [VB](https://github.com/dotnet/roslyn/pull/7540/files#diff-cc9d960f02ff8d02554867434fed4d8cR1), [abstract base class](https://github.com/dotnet/roslyn/pull/7540/files#diff-a3d9e6a718bd711d68d98d4a234dbcaeR1).
- The fix for #7090 (Encapsulate Field should honor `this.` options) is [here](https://github.com/dotnet/roslyn/pull/7540/files#diff-d5796ff4976c53cb7f8c75a959831f72R334) and tests are here: [C#](https://github.com/dotnet/roslyn/pull/7540/files#diff-3700d9dc0f49371327465e65443a7d6bR1202) and [VB](https://github.com/dotnet/roslyn/pull/7540/files#diff-cb2c8c7a3b75808cb4496e1c4d7d4db6R590).  Note that `this.`/`Me.` is always added and then it's the simplifier's job to trim that or not.
- The fix for the VB Encapsulate Field bug where `Me.[Class]` wasn't properly simplified to `[Class]` is [here](https://github.com/dotnet/roslyn/pull/7540/files#diff-521b7192e04645cb866692e23ffda1eeR35) and the test is [here](https://github.com/dotnet/roslyn/pull/7540/files#diff-cb2c8c7a3b75808cb4496e1c4d7d4db6R310).  The corresponding fix for C# is [here](https://github.com/dotnet/roslyn/pull/7540/files#diff-bcb97d922486ad104841ebf01a7ec463R40) and the test is [here](https://github.com/dotnet/roslyn/pull/7540/files#diff-3700d9dc0f49371327465e65443a7d6bR1172).

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest/CodeActions/GenerateFromMembers/GenerateConstructor/GenerateConstructorTests.cs:
```diff
@@ -93,7 +93,7 @@ protected override object CreateCodeRefactoringProvider(Workspace workspace)
 @"class Z { [|public int A { get ; private set ; } public string B { get ; private set ; }|] } ",
 @"class Z { public Z ( int a , string b ) { this . A = a ; this . B = b ; } public int A { get ; private set ; } public string B { get ; private set ; } } ",
-index: 0, options: new Dictionary<OptionKey, object> { { new OptionKey(SimplificationOptions.QualifyMemberAccessWithThisOrMe, "C#"), true } });
+index: 0, options: new Dictionary<OptionKey, object> { { new OptionKey(SimplificationOptions.QualifyPropertyAccess, LanguageNames.CSharp), true } });
```
Please add some sort o helper that makes this options object for you.  i.e. ```Options(QuaifyPropertyAccess, true)```  And then use that all over the place.

DustinCampbell replied:
I think I said that in the last PR too. This is really cumbersome to have to do every time. Definitely a good place for a helper. Of course, if you add the helper, make sure to use it everywhere. :smile:

brettfo replied:
I added a helper for the newly added test cases, this was just updating an already present test, but I'll expand the helper to be more helpful.

brettfo replied:
Done.

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest/Diagnostics/QualifyMemberAccess/QualifyMemberAccessTests.cs:
```diff
@@ -0,0 +1,229 @@
+            return Tuple.Create<DiagnosticAnalyzer, CodeFixProvider>(new CSharpQualifyMemberAccessDiagnosticAnalyzer(), new CSharpQualifyMemberAccessCodeFixProvider());
+        }
+
+        private async Task TestAsyncWithOption(string code, string expected, PerLanguageOption<bool> option)
```
Oh nice.  You have it.  So use the same thing in the above tests as well.

CyrusNajmabadi replied:
This does not need ot be async.  It can just return "TestAsync" directly in its body.

brettfo replied:
Done.

CyrusNajmabadi commented on src/EditorFeatures/VisualBasicTest/CodeActions/EncapsulateField/EncapsulateFieldTests.vb:
```diff
@@ -582,5 +584,29 @@ End Enum
+    End Property
+End Class
+",
+options:=New Dictionary(Of OptionKey, Object)() From {{New OptionKey(SimplificationOptions.QualifyFieldAccess, LanguageNames.VisualBasic), True}})
```
I don't want to see "new Dictionary" anywhere in this CR except in a helper function :)

brettfo replied:
Done.

CyrusNajmabadi commented on src/EditorFeatures/VisualBasicTest/Diagnostics/QualifyMemberAccess/QualifyMemberAccessTests.vb:
```diff
@@ -0,0 +1,207 @@
+            Return Tuple.Create(Of DiagnosticAnalyzer, CodeFixProvider)(New VisualBasicQualifyMemberAccessDiagnosticAnalyzer(), New VisualBasicQualifyMemberAccessCodeFixProvider())
+        End Function
+
+        Private Async Function TestAsyncWithOption(code As String, expected As String, opt As PerLanguageOption(Of Boolean)) As Task
```
same comment about not needing ot be async.

brettfo replied:
Done.

CyrusNajmabadi commented on src/EditorFeatures/VisualBasicTest/Diagnostics/SimplifyTypeNames/SimplifyTypeNamesTests_FixAllTests.vb:
```diff
@@ -288,7 +288,7 @@ End Class]]>
         <Trait(Traits.Feature, Traits.Features.CodeActionsFixAllOccurrences)>
         Public Async Function TestFixAllInSolution_RemoveMe() As Task
-            Dim fixAllActionId = SimplifyTypeNamesCodeFixProvider.GetCodeActionId(IDEDiagnosticIds.SimplifyThisOrMeDiagnosticId, Nothing)
+            Dim fixAllActionId = SimplifyTypeNamesCodeFixProvider.GetCodeActionId(IDEDiagnosticIds.RemoveThisOrMeDiagnosticId, Nothing)
```
Can we call this the "RemoveQualificationDiagnosticId"?

brettfo replied:
Done.

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeFixes/QualifyMemberAccess/CSharpQualifyMemberAccessCodeFixProvider.cs:
```diff
@@ -0,0 +1,29 @@
+    [ExtensionOrder(After = PredefinedCodeFixProviderNames.RemoveUnnecessaryCast)]
+    internal class CSharpQualifyMemberAccessCodeFixProvider : AbstractQualifyMemberAccessCodeFixprovider<SimpleNameSyntax>
+    {
+        protected override string GetId() => CSharpFeaturesResources.AddThisQualification;
```
I would call this "Qualify member access".  It will be clear from the preview that that's done by adding "this.".  I don't like messages/titles that literally state exactly what the code change is.  that feels redundant.

Bonus: by doing this, you won't need another title if we add static qualification.

brettfo replied:
Done.

dpoeschl replied:
When you say "done" you just mean locally, right?

brettfo replied:
At the time of my comment I meant just locally, but I've since pushed the changes to this PR.

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeFixes/QualifyMemberAccess/CSharpQualifyMemberAccessCodeFixProvider.cs:
```diff
@@ -0,0 +1,29 @@
+
+        protected override string GetTitle() => CSharpFeaturesResources.AddThisQualification;
+
+        protected override SyntaxNode GetReplacement(SimpleNameSyntax node)
```
Can't this all be done in a language neutral way by using the SyntaxGenerator?

brettfo replied:
Good catch.  I've updated it to use the syntax generator.  Once I fix my merge conflicts after @jasonmalinowski's huge change, I'll update this PR.

CyrusNajmabadi commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpQualifyMemberAccessDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,30 @@
+
+        protected override bool IsCandidate(SyntaxNode node)
+        {
+            return (node.Parent as MemberAccessExpressionSyntax)?.Expression.Kind() != SyntaxKind.ThisExpression;
```
Very cute :)

CyrusNajmabadi commented on src/Features/CSharp/Portable/EncapsulateField/CSharpEncapsulateFieldService.cs:
```diff
@@ -37,9 +37,15 @@ internal class CSharpEncapsulateFieldService : AbstractEncapsulateFieldService
-                                             .WithTrailingTrivia(declarator.Identifier.TrailingTrivia)
-                                             .WithLeadingTrivia(declarator.Identifier.LeadingTrivia);
+            var escapedName = originalFieldName.EscapeIdentifier();
+            var newIdentifier = SyntaxFactory.Identifier(
```
Does SyntaxFactory.Identifier not already put on elastic trivia?

brettfo replied:
The only overload of `.Identifier()` with separate `text` and `valueText` parameters requires that the leading and trailing trivia be explicitly specified.

CyrusNajmabadi commented on src/Features/CSharp/Portable/EncapsulateField/CSharpEncapsulateFieldService.cs:
```diff
@@ -37,9 +37,15 @@ internal class CSharpEncapsulateFieldService : AbstractEncapsulateFieldService
-            var newIdentifier = SyntaxFactory.Identifier(originalFieldName)
-                                             .WithTrailingTrivia(declarator.Identifier.TrailingTrivia)
-                                             .WithLeadingTrivia(declarator.Identifier.LeadingTrivia);
+            var escapedName = originalFieldName.EscapeIdentifier();
```
Why are we doing this?  Why not get the text as originally written in the declaration of the field?  that way if it uses things like ```\u``` we preserve that.

brettfo replied:
I poked around in the EncapsulateFieldService a bit and everywhere we're using `IFieldSymbol.Name` which always reports "a" instead of "\u0061", so it'll be more than just fixing this location to properly preserve "\u0061".

I'll file a separate bug.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeFixes/PredefinedCodeFixProviderNames.cs:
```diff
@@ -28,6 +28,7 @@ internal static class PredefinedCodeFixProviderNames
         public const string ImplementInterface = "Implement Interface";
         public const string InsertMissingCast = "InsertMissingCast";
         public const string MoveToTopOfFile = "Move To Top Of File";
+        public const string QualifyMemberAccess = "Qualify member access";
```
all these names are terrible.  can you use nameof(QualifyMemberAccess), and fix the rest?

brettfo replied:
I was under the impression that these strings needed to be human-readable because they're set as the name of the code fix provider in [`ExportCodeFixProviderAttribute`](http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/CodeFixes/ExportCodeFixProviderAttribute.cs,49010341106c0e4e).

dpoeschl replied:
Pretty sure the names are only used for extension ordering attributes. If they need to be human-readable then we have a loc bug, but I don't think they do. Please verify the Name isn't used anywhere and use Cyrus's ```nameof``` suggestion or file a loc bug with the appropriate Tenet-* labels.

brettfo replied:
Confirmed that they're only used in ordering attributes.  I've updated all of them to use `nameof()`.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeFixes/Qualify/AbstractQualifyMemberAccessCodeFixprovider.cs:
```diff
@@ -0,0 +1,47 @@
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var model = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            var token = root.FindToken(span.Start, findInsideTrivia: true);
```
Why FindInsideTrivia?

brettfo replied:
Because copypasta :)

I'll fix it.

DustinCampbell replied:
I suppose that would make it work for XML doc comment CREFs. However, do we want it to?

brettfo replied:
Doc comments don't allow `<see cref="this.foo" />` so I'll remove the trivia search.

dpoeschl replied:
But they do allow `Typename.StaticMember`, so you may need to revisit this if/when we support that.

CyrusNajmabadi commented on src/Features/Core/Portable/Diagnostics/Analyzers/IDEDiagnosticIds.cs:
```diff
@@ -6,9 +6,10 @@ internal static class IDEDiagnosticIds
         public const string SimplifyNamesDiagnosticId = "IDE0001";
         public const string SimplifyMemberAccessDiagnosticId = "IDE0002";
-        public const string SimplifyThisOrMeDiagnosticId = "IDE0003";
+        public const string RemoveThisOrMeDiagnosticId = "IDE0003";
```
SimplifyMemberAccessId, QualifyMemberAccessId?

brettfo replied:
Done.

CyrusNajmabadi commented on src/Features/Core/Portable/Diagnostics/Analyzers/QualifyMemberAccessDiagnosticAnalyzerBase.cs:
```diff
@@ -0,0 +1,88 @@
+
+        private OptionSet _lazyDefaultOptionSet;
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
```
can use =>

brettfo replied:
Done.

CyrusNajmabadi commented on src/Features/Core/Portable/Diagnostics/Analyzers/QualifyMemberAccessDiagnosticAnalyzerBase.cs:
```diff
@@ -0,0 +1,88 @@
+
+        private void AnalyzeNode(SyntaxNodeAnalysisContext context)
+        {
+            if (!IsCandidate(context.Node))
```
I have a deep seated belief that this should be doable with IOperation.  And if it can't (because IOperation can't distinguish between "foo" and "this.foo" then we need to fix IOperation.   It should be able to tell us if we're accessing "this" and if that access is implicit or explicit.

DustinCampbell replied:
"deep seated" should be hyphenated.

brettfo replied:
I looked at `IOperation` and I didn't see anything that hinted that it would know the difference between `foo` and `this.foo` and I'm in the opposite camp, that `IOperation` **shouldn't** know the difference because the semantic meaning is the same.  I'll walk down the hall to discuss this further :)

CyrusNajmabadi replied:
I'm meeting @JohnHamby at 3 to discuss this.  Want to come along?  

I personally disagree.  I think it's totally possible to have a single type to represent accessing a member in IOperation.  But that same IOperation can expose if that access was implicit or explicit.  

To me, this is the poster child for the sort of operation you want to do a in language neutral manner.  "Are my accesses qualified?  No.  Does the user want them qualified?  Yes.  Fix them up".  All of that should need no language specific smarts.  IOperation can answer the question, and SyntaxGenerator can provide the fix.

CyrusNajmabadi replied:
Actually, i think what i'm talking about already exists:

```c#
    /// <summary>
    /// Represents a reference to a C# this or VB Me parameter.
    /// </summary>
    public interface IInstanceReferenceExpression : IParameterReferenceExpression
    {
        /// <summary>
        /// Indicates whether the reference is explicit or implicit in source.
        /// </summary>
        bool IsExplicit { get; }
    }
```

So we already have an IOperation type that tells you you're referencing something on the instance.  *and* that node will also tell you if the 'this/me' is explicit or implicit.  This seems like precisely what we want to be using here.  

With this we should not need any language specific code for this feature.

brettfo replied:
Unfortunately `IsExplicit` is [hard-coded to return `true`](http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp/BoundTree/Expression.cs,547).  I've filed #7582 to track this.  In the meantime I'll continue to use the syntactic approach.

CyrusNajmabadi commented on src/Features/Core/Portable/FeaturesResources.resx:
```diff
@@ -670,7 +670,7 @@ Do you want to continue?</value>
     <value>Simplify Member Access</value>
   </data>
-  <data name="SimplifyThisOrMe" xml:space="preserve">
+  <data name="RemoveThisOrMe" xml:space="preserve">
```
I really don't want ot see any C# or VBisms at the neutral layer.

brettfo replied:
Changed to `RemoveQualification` (and updated the actual resource text accordingly.)

DustinCampbell replied:
Me either.

CyrusNajmabadi commented on src/Features/VisualBasic/Portable/Diagnostics/Analyzers/VisualBasicQualifyMemberAccessDiagnosticAnalyzer.vb:
```diff
@@ -0,0 +1,33 @@
+
+        Protected Overrides Function IsCandidate(node As SyntaxNode) As Boolean
+            If TypeOf node.Parent Is MemberAccessExpressionSyntax Then
+                Return If(CType(node.Parent, MemberAccessExpressionSyntax)?.Expression?.Kind(), SyntaxKind.None) <> SyntaxKind.MeExpression
```
do we have any concerns about ```mybase.whatever``` or ```base.whatever```?

brettfo replied:
To me that seems like that would be a different style option.  This PR is just about qualifying member access, not static access or base access.

CyrusNajmabadi replied:
I agree.  However, what i meant was: can we have tests to make sure that we don't mess up in cases where we have 'base' or 'static' or 'pointer' access :)

Right now you're checking just for MemberAccessExpression.  That encompasses a lot.  I want to make sure we don't mess things up when the above returns 'true' when called on 'MyBase.Blah'.

CyrusNajmabadi commented on src/VisualStudio/CSharp/Impl/Options/AutomationObject.cs:
```diff
@@ -445,10 +445,28 @@ public int Style_PreferIntrinsicPredefinedTypeKeywordInMemberAccess
+            set { SetBooleanOption(SimplificationOptions.QualifyFieldAccess, value); }
+        }
+
+        public int Style_QualifyPropertyAccessWithThisOrMe
```
this should be Style_QualifyPropertyAccess

brettfo replied:
Fixed in VB and C#.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/Simplification/SimplificationOptions.cs:
```diff
@@ -45,11 +46,32 @@ public static class SimplificationOptions
+        /// This option says if we should simplify away the <see langword="this"/>. or <see langword="Me"/>. in member access expressions.
         /// </summary>
+        [Obsolete]
         public static PerLanguageOption<bool> QualifyMemberAccessWithThisOrMe { get; } = new PerLanguageOption<bool>(PerLanguageFeatureName, "QualifyMemberAccessWithThisOrMe", defaultValue: false);
```
When was this added?  Can we remove this instead of making it obsolete?

brettfo replied:
It's already in the public API; I wasn't sure if it could be removed or not.

DustinCampbell replied:
Well, you've already made the option no longer work. That's also an API breaking change. Perhaps we should keep this this as some sort of parent option that sets or unsets all of the more-specific options underneath it? In the UI, there'd be this checkbox with the various member checkboxes nested underneath.

dpoeschl replied:
Personally, as far as the UI is concerned, I'd rather pretend the old setting doesn't exist. If someone sets the setting programmatically then we can make that the setting for the 4 specific settings to maintain API compatibility, but for the UI a top-level checkbox with 4 checkboxes underneath feels overkill.

brettfo replied:
I think it should be fully deprecated and no longer used.

@DustinCampbell: I thought about using it as a parent option like you suggested (e.g., setting to `true` or `false` would set/unset all child options), but what would be the expected value if some other extension reads this option?  If all child values are `true`, return `true`; if all child values are `false`, return false, but what about if only some of the child options are `true`?  What do we return then?  We could change it to a `bool?`, but that's still a breaking API change and still doesn't properly convey the information we need.

DustinCampbell replied:
I'm with @dpoeschl that UI is probably OK to drop. However, this is a real API breaking change that we should take through the appropriate channels (e.g. compatibility council).

DustinCampbell commented on src/EditorFeatures/VisualBasicTest/CodeActions/EncapsulateField/EncapsulateFieldTests.vb:
```diff
@@ -305,7 +307,7 @@ Class C
         End Get
         Set(value As String)
-            Me.Class = value
+            [Class] = value
```
Could you add another test that performs this particular escaping scenario with `QualifyFieldAccess` set to true?

brettfo replied:
Good catch, will do.

DustinCampbell commented on src/Features/Core/Portable/CodeFixes/Qualify/AbstractQualifyMemberAccessCodeFixprovider.cs:
```diff
@@ -0,0 +1,47 @@
+
+            var codeAction = new QualifyMemberAccessCodeAction(
+                GetTitle(),
+                c => document.ReplaceNodeAsync(node, GetReplacement(node), c),
```
Shouldn't a Simplifier.Annotation get applied to the new node? It looked like this didn't happen in the override of `GetReplacement`

DustinCampbell replied:
Note, I'm not sure if SyntaxGenerator does this automatically.

brettfo replied:
I can go either way on this.  The only thing that `GetReplacement()` does is add a `this.` prefix, and by definition of this style rule, that can't get simplified, but I can also argue that if the simplifier ever gets more complicated (see what I did there?) that this would be necessary.

Would there be a noticeable perf hit in adding a simplifier annotation that we know will result in a noop?

DustinCampbell replied:
There wouldn't really be a perf hit, but yes, it'd be essentially a no-op. So, for the purpose of this Code Action, it's probably fine.

DustinCampbell commented on src/Features/VisualBasic/Portable/Diagnostics/Analyzers/VisualBasicQualifyMemberAccessDiagnosticAnalyzer.vb:
```diff
@@ -0,0 +1,33 @@
+        Protected Overrides Function IsCandidate(node As SyntaxNode) As Boolean
+            If TypeOf node.Parent Is MemberAccessExpressionSyntax Then
+                Return If(CType(node.Parent, MemberAccessExpressionSyntax)?.Expression?.Kind(), SyntaxKind.None) <> SyntaxKind.MeExpression
+            End If
```
I think this whole method can be reduced to:

```VB
Protected Overrides Function IsCandidate(node as SyntaxNode) As Boolean
    Return If(TryCast(node.Parent, MemberAccessExpressionSyntax)?.Expression?.Kind(), SyntaxKind.None) <> SyntaxKind.MeExpression
End Function

brettfo replied:
Thanks, I forgot about `TryCast()`

dpoeschl commented:
@brettfo Can you link to the closed changes when they're available? OptionPagesDTE.xml is the only test failure.

dpoeschl commented on src/Features/CSharp/Portable/CodeFixes/QualifyMemberAccess/CSharpQualifyMemberAccessCodeFixProvider.cs:
```diff
@@ -0,0 +1,29 @@
+namespace Microsoft.CodeAnalysis.CSharp.CodeFixes.QualifyMemberAccess
+{
+    [ExportCodeFixProvider(LanguageNames.CSharp, Name = PredefinedCodeFixProviderNames.QualifyMemberAccess), Shared]
+    [ExtensionOrder(After = PredefinedCodeFixProviderNames.RemoveUnnecessaryCast)]
```
Why do you want to be after RemoveUnnecessaryCast?

brettfo replied:
Because that's what SimplifyTypeName does and ultimately they're the same code fixes; just that one is the compliment to the other.

brettfo commented:
@dpoeschl: I linked to them in an earlier comment that got lost: dotnet/roslyn-internal#430

CyrusNajmabadi commented on src/Features/Core/Portable/CodeFixes/PredefinedCodeFixProviderNames.cs:
```diff
@@ -4,37 +4,38 @@ namespace Microsoft.CodeAnalysis.CodeFixes
+        public const string RenameTracking = nameof(RenameTracking);
+        public const string SimplifyNames = nameof(SimplifyNames);
+        public const string SpellCheck = nameof(SpellCheck);
+        public const string Suppression = nameof(Suppression);
```
Yaay

balajikris commented on src/Workspaces/Core/Portable/Simplification/QualifyMemberAccessCodeAction.cs:
```diff
@@ -0,0 +1,17 @@
+
+namespace Microsoft.CodeAnalysis.Simplification
+{
+    internal class QualifyMemberAccessCodeAction : CodeAction.DocumentChangeAction
```
Note: This isn't accessed from anywhere but AbstractQualifyMemberAccessCodeFixprovider.cs, so it could be a private type within `AbstractQualifyMemberAccessCodeFixprovider`

brettfo replied:
Good catch.  I'll move the type.

brettfo commented:
I believe I've now addressed all feedback with the glaring exception of discussing the breaking API change with @DustinCampbell.

One notable change with the most recent commit is that qualifying methods and events currently doesn't work due to limitations/bugs in the compiler around generating `IOperations`s.  These exist as issues #7584 and #7587.

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest/CodeActions/GenerateFromMembers/GenerateConstructor/GenerateConstructorTests.cs:
```diff
@@ -93,7 +93,7 @@ protected override object CreateCodeRefactoringProvider(Workspace workspace)
 @"class Z { [|public int A { get ; private set ; } public string B { get ; private set ; }|] } ",
 @"class Z { public Z ( int a , string b ) { this . A = a ; this . B = b ; } public int A { get ; private set ; } public string B { get ; private set ; } } ",
-index: 0, options: new Dictionary<OptionKey, object> { { new OptionKey(SimplificationOptions.QualifyMemberAccessWithThisOrMe, "C#"), true } });
+index: 0, options: Option(SimplificationOptions.QualifyPropertyAccess, true));
```
thanks!

CyrusNajmabadi commented on src/Features/Core/Portable/Diagnostics/Analyzers/QualifyMemberAccessDiagnosticAnalyzerBase.cs:
```diff
@@ -0,0 +1,110 @@
+        private void AnalyzeOperation(OperationAnalysisContext context)
+        {
+            var memberReference = (IMemberReferenceExpression)context.Operation;
+            if (memberReference.Instance == null)
```
Comment for completeness what this means.  i.e. "this is a static member reference, so we don't want to qualify with this/me".

brettfo replied:
Done.

CyrusNajmabadi commented on src/Features/Core/Portable/Diagnostics/Analyzers/QualifyMemberAccessDiagnosticAnalyzerBase.cs:
```diff
@@ -0,0 +1,110 @@
+                return;
+            }
+
+            if (IsAlreadyQualifiedMemberAccess(memberReference.Instance.Syntax))
```
what happens if i have "class C { void Foo(C c) { c.Foo(); } }"  Do you do the right thing here?  First, we will have an instance.  But the instance won't be 'this'.  But hte member symbol's containing symbol will be the instance symbol's type.  But we don't want to qualify this.

brettfo replied:
Good catch.  I've added `if (memberReference.Instance.Kind != OperationKind.InstanceReferenceExpression) { return; }` because in your case the kind is `ParameterReference` or it could be `LocalReference`, etc.

I've also added tests.  (I'll push the changes shortly to this PR.)

CyrusNajmabadi commented on src/Features/Core/Portable/Diagnostics/Analyzers/QualifyMemberAccessDiagnosticAnalyzerBase.cs:
```diff
@@ -0,0 +1,110 @@
+
+        private static ISymbol GetSymbol(IMemberReferenceExpression memberReferenceExpression)
+        {
+            switch (memberReferenceExpression.Kind)
```
Put a comment in that this can be simplified once IMemberReferenceExpression just exposes 'Member'

brettfo replied:
Done.

CyrusNajmabadi commented on src/Features/Core/Portable/Diagnostics/Analyzers/QualifyMemberAccessDiagnosticAnalyzerBase.cs:
```diff
@@ -0,0 +1,110 @@
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(s_descriptorQualifyMemberAccess);
+
+        protected abstract ImmutableArray<TLanguageKindEnum> GetSupportedSyntaxKinds();
```
Do you use SupportedSyntaxKinds anymore?

brettfo replied:
Good catch.  I've removed it.

CyrusNajmabadi commented on src/Features/VisualBasic/Portable/CodeFixes/QualifyMemberAccess/VisualBasicQualifyMemberAccessCodeFixProvider.vb:
```diff
@@ -0,0 +1,13 @@
+
+    <ExportCodeFixProvider(LanguageNames.VisualBasic, Name:=PredefinedCodeFixProviderNames.QualifyMemberAccess), [Shared]>
+    <ExtensionOrder(After:=PredefinedCodeFixProviderNames.RemoveUnnecessaryCast)>
+    Friend Class VisualBasicQualifyMemberAccessCodeFixProvider
```
Is this even necessary anymore?  Is it possible to just have the base type and have two ExportCodeFixProviders on it (one for VB and one for C#)?  If that's not possible, then it's fine to have this type.  I'm just ecstatic that it's so empty.

brettfo replied:
We currently still require this for the `IsAlreadyQualifiedMemberAccess()` override, but once #7582 is fixed then C#/VB can be combined.

CyrusNajmabadi commented on src/Features/VisualBasic/Portable/Diagnostics/Analyzers/VisualBasicQualifyMemberAccessDiagnosticAnalyzer.vb:
```diff
@@ -0,0 +1,24 @@
+
+        Private Shared ReadOnly s_kindsOfInterest As ImmutableArray(Of SyntaxKind) = ImmutableArray.Create(SyntaxKind.IdentifierName)
+
+        Protected Overrides Function GetSupportedSyntaxKinds() As ImmutableArray(Of SyntaxKind)
```
Why are these kinds needed?

brettfo replied:
They're not anymore.  Removed.

CyrusNajmabadi commented:
I love how much simpler this is.  :+1:

CyrusNajmabadi commented:
Sounds like @JohnHamby is on point for working through the IOperation issues.  This will be a great example of how simple it is to write a great analyzer/fix combo in a language neutral way.  Great job!

jasonmalinowski commented on src/EditorFeatures/Test/CodeActions/AbstractCodeActionOrUserDiagnosticTest.cs:
```diff
@@ -339,5 +338,10 @@ protected static IList<CodeAction> FlattenActions(IEnumerable<CodeAction> codeAc
             return codeActions?.SelectMany(a => a.HasCodeActions ? a.GetCodeActions().ToArray() : new[] { a }).ToList();
         }
+
+        protected IDictionary<OptionKey, object> Option(PerLanguageOption<bool> option, bool value)
```
Nice.

jasonmalinowski commented on src/EditorFeatures/VisualBasicTest/Diagnostics/QualifyMemberAccess/QualifyMemberAccessTests.vb:
```diff
@@ -0,0 +1,358 @@
+
+        <WorkItem(7065, "https://github.com/dotnet/roslyn/issues/7065")>
+        <Fact, Trait(Traits.Feature, Traits.Features.CodeActionsQualifyMemberAccess)>
+        Public Async Function QualifyFieldAccess_RHS() As Task
```
Do we have some tests with property or event backing fields?

jasonmalinowski replied:
Also consider: accessing statics through Me (valid in VB!) and VB modules. I don't think you have any bugs here but it's probably reasonable.

brettfo replied:
Good idea.  I've added a test for qualifying the backing field on an auto-property when `QualifyFieldAccess` is true.  I opted not to try to qualify the backing field if only Qualify**Property**Access is true because any developer explicitly accessing the backing field presumably knows what they're doing and would be confused to see a ``Me.`' qualification being suggested.

That, and it would make this feature needlessly complex while only adding minimal value for 2 people.

I've also added a test for accessing `Shared` (not `static`; what do you think this is, C#?) with a '`Me.`' qualifier.  VB doesn't allow '`Me.`' in modules so that's been skipped.

I'll be adding the commit to this PR with the tests shortly.

jasonmalinowski commented on src/Workspaces/Core/Portable/Simplification/SimplificationOptions.cs:
```diff
@@ -45,11 +46,32 @@ public static class SimplificationOptions
 
         /// <summary>
-        /// This option says if we should simplify away the this. or Me. in member access expression
+        /// This option says if we should simplify away the <see langword="this"/>. or <see langword="Me"/>. in member access expressions.
```
If you're going to make this less terrible, we probably should remove the . after this so it's grammatically correct.

brettfo replied:
Done.

jasonmalinowski commented on src/Features/Core/Portable/Diagnostics/Analyzers/QualifyMemberAccessDiagnosticAnalyzerBase.cs:
```diff
@@ -0,0 +1,120 @@
+            }
+        }
+
+        // TODO: this method can be removed once `IMemberReferenceExpression` exposes a `Member` property
```
Is there a bug tracking this, and some pointer to this code? Or can you just implement this as a part of this PR?

brettfo replied:
I've filed #7608 to track this and updated this comment with that bug number.

jasonmalinowski commented on src/Features/Core/Portable/Diagnostics/Analyzers/QualifyMemberAccessDiagnosticAnalyzerBase.cs:
```diff
@@ -0,0 +1,120 @@
+                return workspaceOptions.Workspace.Options;
+            }
+
+            if (_lazyDefaultOptionSet == null)
```
Should this just be a helper we have somewhere?

brettfo replied:
Done.

jasonmalinowski commented on src/Features/Core/Portable/Diagnostics/Analyzers/QualifyMemberAccessDiagnosticAnalyzerBase.cs:
```diff
@@ -0,0 +1,120 @@
+            }
+
+            // if the reference is to a base type
+            if (memberSymbol.ContainingSymbol != memberReference.Instance.ResultType)
```
This seems odd. Why would I not want `this.SomeMethodInByBase()`? And in general, do we have enough tests to make sure no sort of `base` references break anything here?

brettfo replied:
`base.` and `MyBase.` references won't ever get this far because they're taken care of on line 43 with `if (memberReference.Instance.Kind != OperationKind.InstanceReferenceExpression) { return }` (and there's a comment there explaining this.)

I've experimented with removing this check (line 71) and we appropriately flag `SomeBaseProperty = 1` as convertable to `Me.SomeBaseProperty = 1`, but since we don't yet have an analyzer/code fix/simplification rule for `base.`/`MyBase.` then you can get into a weird scenario of `MyBase.Property` gets flagged as being able to be simplified to `Property`, but then that's immediately flagged as being able to be qualified with `Me.`.

Ultimately, I'm on the fence about this one and don't care either way.  If you feel strongly about qualifying base access with `Me.` then I can easily remove this check.

jasonmalinowski replied:
To me this feels like a bug we're immediately going to get. If I say all things should be qualified, all things should be qualified. Implicit references to base types count as "all".

brettfo replied:
Fair enough.  I'll remove the check and update the tests.

jasonmalinowski commented:
Overall :+1:, but the explicit blocking of `this` qualification on references to base type members seems surprising to me. Otherwise just some nitpicks around more possibilities for tests.

brettfo commented:
Feedback addressed, but closing in favor of #8435 against `future`.
